### PR TITLE
[KUBOS-350] CSP Modifications

### PR DESCRIPTION
**Pull request by plauche**
* Switched tcp socket interface to use SCTP for more reliable message delivery
* Introduced cbor serialization for tcp socket communication
* Added some termination/cleanup functions
* Modified csp build structure to allow for yotta building of tests
* Added unit test and integration tests to exercise tcp socket interface

### CODE REVIEW

File: libcsp/src/CMakeLists.txt [hunk 1]
```diff
@@ -12,13 +12,13 @@ option (XTEA "" OFF)
...
-option (VERBOSE "" OFF)
+option (CSP_VERBOSE "" ON)
 option (IF_I2C "" OFF)
 option (IF_KISS "" OFF)
 option (IF_CAN "" OFF)
-option (IF_SOCKET "" OFF)
+option (IF_SOCKET "" ON)
 option (IF_ZMQHUB "" OFF)
-option (CSP_DEBUG "" OFF)
+option (CSP_DEBUG "" ON)
```
Comment (CatFreed) [inline libcsp/src/CMakeLists.txt]: Are you leaving CSP_DEBUG on on purpose?

File: libcsp/src/csp_conn.c [hunk 2]
```diff
@@ -179,6 +179,27 @@ int csp_conn_init(void) {
...
+		csp_log_error("Failed to remove sport semaphore");
+	}
+
+	int i, prio;
+	for (i = 0; i < CSP_CONN_MAX; i++) {
+		for (prio = 0; prio < CSP_RX_QUEUES; prio++)
+			csp_queue_remove(arr_conn[i].rx_queue[prio]);
+            
+		if (csp_mutex_remove(&arr_conn[i].lock) != CSP_MUTEX_OK) {
+			csp_log_error("Failed to remove connection lock");
```
Comment (CatFreed) [inline libcsp/src/csp_conn.c]: Could it be useful to mention which connection lock couldn't be removed?

File: libcsp/src/csp_qfifo.c [hunk 3]
```diff
@@ -50,6 +50,23 @@ int csp_qfifo_init(void) {
 
 }
 
+void csp_qfifo_terminate(void) {
+    int prio;
+
+	/* Create router fifos for each priority */
```
Comment (CatFreed) [inline libcsp/src/csp_qfifo.c]: Nit: Create->remove

File: libcsp/src/csp_qfifo.c [hunk 4]
```diff
@@ -50,6 +50,23 @@ int csp_qfifo_init(void) {
...
+	for (prio = 0; prio < CSP_ROUTE_FIFOS; prio++) {
+		if (qfifo[prio] != NULL) {
+			csp_queue_remove(qfifo[prio]);
+		}
+	}
+
+#ifdef CSP_USE_QOS
+	/* Create QoS fifo notification queue */
```
Comment (CatFreed) [inline libcsp/src/csp_qfifo.c]: Nit: create->remove

File: libcsp/src/drivers/socket/packet.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+{
+    CborEncoder encoder, container;
+    CborError err;
+
+    if (buffer == NULL)
+        return -1;
+    cbor_encoder_init(&encoder, buffer, SOCKET_BUFFER_SIZE, 0);
+    err = cbor_encoder_create_map(&encoder, &container, 3);
+    if (err > 0)
+        return - err;
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/packet.c]: nit: spacing

File: libcsp/src/drivers/socket/packet.c [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+int cbor_encode_csp_packet(csp_packet_t * packet, uint8_t * buffer)
+{
+    CborEncoder encoder, container;
+    CborError err;
+
+    if (buffer == NULL)
+        return -1;
+    cbor_encoder_init(&encoder, buffer, SOCKET_BUFFER_SIZE, 0);
+    err = cbor_encoder_create_map(&encoder, &container, 3);
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/packet.c]: Where does the '3' value come from?

File: libcsp/src/drivers/socket/packet.c [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    int length = 0;
+    char * buf = NULL;
+
+    if ((buffer == NULL) || (packet == NULL))
+        return false;
+
+    CborError err = cbor_parser_init(buffer, buffer_size, 0, &parser, &map);
+    if (err || !cbor_value_is_map(&map)) {
+        return false;
+    }
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/packet.c]: styling: You don't use brackets for your other one-line if statements.

Personally, I'd prefer that all if statements use brackets (and it's what in our (very limited) [coding standards doc](http://docs.kubos.co/0.2.0/md_docs_kubos-standards.html))

File: libcsp/src/drivers/socket/packet.c [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    cbor_value_calculate_string_length(&element, &len);
+
+    size_t byte_size;
+    err = cbor_value_dup_text_string(&element, &buf, &byte_size, &element);
+    if (err) {
+        csp_log_error("Error parsing text string %d\r\n", err);
+       free(buf);
+        return false;
+    } else {
+        csp_log_info("Got dat text string %d %d\r\n", byte_size, packet->length);
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/packet.c]: nit: should be 'data'?

File: libcsp/src/drivers/socket/socket_linux.c [hunk 9]
```diff
@@ -51,11 +52,15 @@ static int socket_server_init(csp_socket_handle_t * socket_iface, uint16_t port)
     static bool server_init = false;
 
     if (server_init == false) {
-        server_socket = socket(AF_INET, SOCK_STREAM, 0);
+        server_socket = socket(AF_INET, SOCK_STREAM, IPPROTO_SCTP);
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/socket_linux.c]: This probably means that you want SCTP to be enabled on the iOBC...

File: libcsp/src/drivers/socket/socket_linux.c [hunk 10]
```diff
@@ -65,19 +70,21 @@ static int socket_server_init(csp_socket_handle_t * socket_iface, uint16_t port)
...
 
-        listen(server_socket, 3);
+        listen(server_socket, SERVER_MAX_CONNECTIONS);
         c = sizeof(struct sockaddr_in);
         server_init = true;
     }
 
     csp_log_info("Wait to accept\n");
-    socket_handle = accept(server_socket, (struct sockaddr *)&client, (socklen_t*)&c);
+    // socket_handle = accept(server_socket, (struct sockaddr *)&client, (socklen_t*)&c);
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/socket_linux.c]: nit: remove unused code

File: libcsp/src/drivers/socket/socket_linux.c [hunk 11]
```diff
@@ -119,6 +127,24 @@ int socket_init(csp_socket_handle_t * socket_iface, uint8_t mode, uint16_t port)
     }
 }
 
+int socket_close(csp_socket_handle_t * socket_driver) {
+    if (socket_driver == NULL) {
+        return CSP_ERR_DRIVER;
+    }
+
+    socket_driver->is_active = false;
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/socket_linux.c]: Will anything weird happen if socket_close is called, but the socket_driver isn't active?

File: libcsp/src/interfaces/csp_if_socket.c [hunk 12]
```diff
@@ -38,57 +38,87 @@
...
+
+    csp_log_info("Starting socket rx thread\r\n");
 
     if (param == NULL) {
         csp_log_error("No socket param found\r\n");
-        return CSP_TASK_RETURN;
+        csp_thread_exit();
     }
-    socket_interface = *((csp_iface_t*)param);
+    // socket_interface = *((csp_iface_t*)param);
```
Comment (CatFreed) [inline libcsp/src/interfaces/csp_if_socket.c]: nit: unused code

File: libcsp/src/interfaces/csp_if_socket.c [hunk 13]
```diff
@@ -38,57 +38,87 @@
...
+            {
+                csp_log_info("Got valid csp packet\r\n");
+                csp_log_info("csp_socket_rx got packet length %d\r\n", packet->length);
+                csp_new_packet(packet, &socket_interface, NULL);
+                packet = csp_buffer_get(SOCKET_BUFFER_SIZE);
+                if (packet == NULL) {
+                    csp_log_error("Out of packet buffers\r\n");
+                    break;
+                }
+            }
```
Comment (CatFreed) [inline libcsp/src/interfaces/csp_if_socket.c]: Worth logging invalid packets received?

File: libcsp/src/drivers/socket/packet.c (hunk 6 above)
Comment (plauche) [inline libcsp/src/drivers/socket/packet.c]: That parameter indicates the number of key/value pairs which will be inserted into the map. In this case we just have the length, id, and data values.

Comment (CatFreed) [inline libcsp/src/drivers/socket/packet.c]: Cool, makes sense.  Is there any chance that new pairs would be added to this in the future?  If so, it'd be worth it to document this

File: libcsp/src/drivers/socket/socket_linux.c (hunk 11 above)
Comment (plauche) [inline libcsp/src/drivers/socket/socket_linux.c]: Hmm good question. There is probably no need to double close if it's not active.

File: libcsp/src/drivers/socket/socket_linux.c (hunk 9 above)
Comment (plauche) [inline libcsp/src/drivers/socket/socket_linux.c]: Haha yes...

File: libcsp/src/csp_conn.c [hunk 14]
```diff
@@ -179,6 +179,27 @@ int csp_conn_init(void) {
...
+void csp_conn_terminate(void) {
+
+    if (csp_bin_sem_remove(&sport_lock) != CSP_SEMAPHORE_OK) {
+		csp_log_error("Failed to remove sport semaphore");
+	}
+
+	int i, prio;
+	for (i = 0; i < CSP_CONN_MAX; i++) {
+		for (prio = 0; prio < CSP_RX_QUEUES; prio++)
+			csp_queue_remove(arr_conn[i].rx_queue[prio]);
```
Comment (CatFreed) [inline libcsp/src/csp_conn.c]: nit: brackets

File: libcsp/src/csp_port.c [hunk 15]
```diff
@@ -100,4 +100,12 @@ int csp_bind(csp_socket_t * socket, uint8_t port) {
 
 }
 
+void csp_close_socket(csp_socket_t * socket) {
+    if (socket == NULL)
+        return;
```
Comment (CatFreed) [inline libcsp/src/csp_port.c]: nit: brackets

File: libcsp/src/csp_conn.c (hunk 14 above)
Comment (plauche) [inline libcsp/src/csp_conn.c]: Yes...but I am copying the style from this file, see lines 151-153 :)

Comment (CatFreed) [inline libcsp/src/csp_conn.c]: Yeah, okay

**Pull request merged**